python/lsst/daf/butler/direct_butler/_direct_butler.py

```diff
@@ -1618,25 +1630,30 @@ def ingest_zip(self, zip_file: ResourcePathExpression, transfer: str = "auto") -
             if registered:
                 _LOG.verbose("Created RUN collection %s as part of zip ingest", run)
 
-        # Do not need expanded dataset refs so can ignore the return value.
-        self._ingest_file_datasets(datasets)
+        progress = Progress("lsst.daf.butler.Butler.ingest", level=VERBOSE)
+        import_info = self._prepare_ingest_file_datasets(
+            datasets, progress, dry_run=dry_run, transfer_dimensions=transfer_dimensions
+        )
```

Comment:
Should this be done in the same transaction below? I think it modifies the database, and if transaction below fails, you will end up with partial modifications.

Reply:
This was part of the reorganization of the transfer_from code so it can be called from ingest. The prepare step cannot be in a transaction because it creates a dataset type and that is not allowed to be in a transaction. No other database changes happen. It's one of the reasons why it's a distinct piece of code. The dimension record writing and file ingest are in the transaction.